drivers: icm42688: pin9 function

ymleung314 opened the pull request:
PIN9 of an ICM42688 can be configured as an interrupt output, external clock input or frame sync output. Pin function can now be set via a sensor attribute.

According to the datasheet, FIFO packet timestamps are scaled with respect to the external clock frequency. I have added a field to the decoder buffer header. It may break existing codebase where the mempool is not large enough for the new header struct.

ymleung314 committed:
drivers: icm42688: pin9 function

PIN9 of an ICM42688 can be configured as an interrupt output, external
clock input or frame sync output. Pin function can now be set via a sensor
attribute.

Signed-off-by: Yau-ming Leung <ymleung314@gmail.com>

ymleung314 commented:
Rebased onto main to fix the irrelevant but failed checks. No more changes unless requested to. Please review.

ymleung314 commented:
DNM. RTC frequency is default to 0 now and will cause divide by zero. Fix soon.

ymleung314 commented:
Fixed, please review.

MaureenHelm commented on drivers/sensor/tdk/icm42688/icm42688_common.c:
```diff
@@ -148,6 +149,47 @@ int icm42688_configure(const struct device *dev, struct icm42688_cfg *cfg)
+		return -EINVAL;
+	}
+
+	k_busy_wait(50);
```
Can you sleep instead?

ymleung314 replied:
Probably, I thought the wait duration is so short that I may just busy wait. Also, the datasheet didn't require waiting, I added it purely as a defensive measure. Same for the other busy waits.

afontaine-invn replied:
I confirmed there is no need to have such wait to access register bank in this chip. And no need also at the end of the procedure. It works fine on other drivers' implementation without delays.

MaureenHelm commented on drivers/sensor/tdk/icm42688/icm42688_common.c:
```diff
@@ -148,6 +149,47 @@ int icm42688_configure(const struct device *dev, struct icm42688_cfg *cfg)
+		return -EINVAL;
+	}
+
+	k_busy_wait(50);
```
Can you sleep instead?

MaureenHelm commented on drivers/sensor/tdk/icm42688/icm42688_common.c:
```diff
@@ -148,6 +149,47 @@ int icm42688_configure(const struct device *dev, struct icm42688_cfg *cfg)
+		return -EINVAL;
+	}
+
+	k_busy_wait(250);
```
Can you sleep instead?

MaureenHelm commented on drivers/sensor/tdk/icm42688/icm42688_decoder.c:
```diff
@@ -290,7 +290,7 @@ static uint32_t accel_period_ns[] = {
 	[ICM42688_DT_ACCEL_ODR_3_125] = UINT32_C(10000000000000) / 31250,
 	[ICM42688_DT_ACCEL_ODR_6_25] = UINT32_C(10000000000000) / 62500,
-	[ICM42688_DT_ACCEL_ODR_12_5] = UINT32_C(10000000000000) / 12500,
+	[ICM42688_DT_ACCEL_ODR_12_5] = UINT32_C(10000000000000) / 125000,
```
Is this related to `drivers: icm42688: pin9 function`? It looks like an unrelated bug fix

ymleung314 replied:
Not related, indeed a separate bug fix, you think I should open another PR for that?

MaureenHelm replied:
Don't need to create another PR, just move it to a new commit in this PR.

MaureenHelm commented on drivers/sensor/tdk/icm42688/icm42688_decoder.c:
```diff
@@ -305,7 +305,7 @@ static uint32_t accel_period_ns[] = {
 
 static uint32_t gyro_period_ns[] = {
-	[ICM42688_DT_GYRO_ODR_12_5] = UINT32_C(10000000000000) / 12500,
+	[ICM42688_DT_GYRO_ODR_12_5] = UINT32_C(10000000000000) / 125000,
```
Is this related to `drivers: icm42688: pin9 function`? It looks like an unrelated bug fix

ymleung314 committed:
drivers: icm42688: fix accel and gyro frequency in period calculation

When calculating accel and gyro period for 12.5Hz, the frequency value is
incorrect. Updated to correct value.

Signed-off-by: Yau-ming Leung <ymleung314@gmail.com>

ymleung314 commented:
Anyway, moved the 12.5 Hz fix to its own commit and removed the waits. Please review again.

teburd reviewed (commented):
seems ok to me but @bperseghetti or @afontaine-invn should review
